Mathlib/Algebra/Module/Injective.lean

```diff
@@ -56,35 +57,32 @@ map to `Q`, i.e. in the following diagram, if `f` is injective then there is an
   ```
 -/
 class Module.Injective : Prop where
-  out : ∀ (X Y : TypeMax.{v,u}) [AddCommGroup X] [AddCommGroup Y] [Module R X] [Module R Y]
+  out : ∀ ⦃X Y : Type v⦄ [AddCommGroup X] [AddCommGroup Y] [Module R X] [Module R Y]
     (f : X →ₗ[R] Y) (_ : Function.Injective f) (g : X →ₗ[R] Q),
     ∃ h : Y →ₗ[R] Q, ∀ x, h (f x) = g x
```

Comment:
I have concerns about this aberration from the design of Module.Projective. There we use a test object in a higher universe (a free module generated by elements of the module): the advantage is we get the lifting property in arbitrary universes [Module.projective_lifting_property](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Projective.html#Module.projective_lifting_property), but the cost is that [IsProjective.iff_projective](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Category/ModuleCat/Projective.html#IsProjective.iff_projective) (what a weird name!) only holds in module universes no lower than the ring.

Your [solution 1](https://github.com/leanprover-community/mathlib4/pull/8905#discussion_r1421608548) would be conforming to a similar design. But now you're adopting a definition of Module.Injective that uses only test objects in the same universe, so you only have the extension property in the same universe but not the higher universes, but the good thing is that [Module.injective_iff_injective_object](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Module/Injective.html#Module.injective_iff_injective_object) holds in any universe. (By the way, it's probably more convenient to state the fully universe-polymorphic extension property assuming Module.Baer (), which will be easier to use than Category.Injective of the ULifts.)

I think the Projective design is reasonable because people are probably most interested in the full lifting property rather than the version restricted to one universe, but I can't say your choice is definitely worse. Of course, if we could prove the extension property in the same universe implies the fully-polymorphic one, we won't need to worry about this. I have some ideas about this: firstly `I → Q` in the Baer condition can be replaced by an injective linear map `I/ker → Q`, so that `#(I/ker) ≤ #Q`, but it's not necessary that `#(R/ker) ≤ #Q`. Secondly maybe we can somehow "replace" `R` by `R/ann(Q)` where `ann(Q)` is the two sided ideal that sends every element of Q to 0, but I can't quite obtain an injective linear map `I/J → R/ann(Q)` such that `I → Q` factors through `I/J`; that requires `I ⊓ ann(Q) ≤ J` and `J ≤ ker` but we only have `ann(Q) * I ≤ ker`. Thirdly I'm now looking into injective hulls, and hopefully something has been said about their cardinalities.

Reply:
```lean

universe uR uM uP uP'

variable [UnivLE.{uR, uM}]
variable (R : Type uR) [Ring R]
variable (M : Type uM) [AddCommGroup M] [Module R M] [inj : Module.Injective R M]
variable (P : Type uP) [AddCommGroup P] [Module R P]
variable (P' : Type uP') [AddCommGroup P'] [Module R P']

lemma Module.Injective.lifting_property
    (f : P →ₗ[R] P') (hf : Function.Injective f)
    (g : P →ₗ[R] M) : ∃ h : P' →ₗ[R] M, h ∘ₗ f = g := by
  have inj' : Module.Injective R (ULift.{max uM uP uP'} M) :=
    Module.ulift_injective_of_injective.{uR, uM, max uM uP uP'} R inj
  obtain ⟨h, H⟩ := inj'.out
    (ULift.moduleEquiv.symm.toLinearMap ∘ₗ f ∘ₗ ULift.moduleEquiv.toLinearMap :
      ULift.{max uM uP uP'} P →ₗ[R] ULift.{max uM uP uP'} P')
    (ULift.moduleEquiv.symm.injective.comp (hf.comp ULift.moduleEquiv.injective))
    (ULift.moduleEquiv.symm.toLinearMap ∘ₗ g ∘ₗ ULift.moduleEquiv.toLinearMap)
  refine ⟨ULift.moduleEquiv.toLinearMap ∘ₗ h ∘ₗ ULift.moduleEquiv.symm.toLinearMap, ?_⟩
  ext x
  specialize H (ULift.up x)
  simp only [LinearMap.coe_comp, LinearEquiv.coe_coe, Function.comp_apply, ULift.moduleEquiv_apply,
    ULift.moduleEquiv_symm_apply] at H ⊢
  rw [H]
```
This is the best result in terms of universe polymorphism I can get using `UnivLE`

Reply:
I didn't quite follow your proof, which linear map are we taking `ker`

Reply:
Yeah that's what I have in mind. But it should be called extension_property not lifting. Also, you could prove Module.Baer.extension_property without UnivLE.

Reply:
`Module.Baer.extension_property` has been added (fully universally polymorphic, maybe like you suggested, `Module.Baer` should be the definitoin of `Module.Injective`

Reply:
That's great!

> I didn't quite follow your proof, which linear map are we taking ker

Just saw this. I mean the kernel of I → Q. You can replace `I → R` by `I/ker → R/ker` but you don't have info about the cardinality of `R/ker`.

Anyway I'm now convinced whether we can test Q with objects in the same universe is equivalent to whether Q's R-injective hull (or rather its cardinality) lies in the same universe. So I really need to look into the construction of injective hulls ... (On the dual side, a projective cover may not exist, but a flat cover always does!)

Reply:
Update: I'm almost sure that being an injective module in a universe lower than the ring's universe doesn't imply being injective in the ring's universe (equivalently, Module.Baer). For details, see [this comment](https://github.com/leanprover-community/mathlib4/pull/9120/files#r1509525127).